### DISCUSSION TITLE
Tasmota energy with Shelly EM on consumption and production

### DISCUSSION THREAD

**Discussion by abroeders** [arendst/Tasmota#15315] [Q&A] [answered]
Hi all,
I use a shelly EM with latest tasmota firmware 11.0.0 , I use 1 channel for the production of my PV installation. Other channel I use for my consumption. 
I expected the production te be negative. I already tried inverting the direction of the transformer but this stays the same.

What is the total KWh with this Shelly module? Does it add up the two channels? I would like it to be daily positive if consumption > production and a negative value if production > consumption.

Is this possible with any option to set this?

**Comment by sfromis** [comment]
As AC has no "direction", when seen from the perspective of the clamp, I'd not expect it to able to detect what's what, and not apply any negative sign. Thus, the logical conclusion would be that "total" is the sum of the two positive channel values. You can do postprocessing in your backend, rules, or whatever.

**Comment by arendst** [comment, marked as answer]
As with all energy monitors there is an option to show total energy per phase or accumulated. Don't know which though.

3em can also show negative values.

Make sure to us latest development version.

**Reply by abroeders** [reply]
I have the em and not the 3em. Do you mean an option to substract the two channels in stead of sum them? That would fix my issue.

**Reply by abroeders** [reply]
I found something:
SetOption129 -> setting 1 splits energyusage in 2 channels in stead of adding them together. This way i can combine them on my backend and substract.

Suggestion: Would it be possible to add a SetOption129 with setting 2 that substracts both channels. So a value of:
0 - Adds the channels (default)
1 - seperate reports
2 - substract the channels

**Reply by sfromis** [reply]
As those SetOption values are a binary on/off, it would have to be done another way.

**Reply by abroeders** [reply]
Other option would be to use SetOption129 0 (sum the channels) and have 1 channel made negative (value = 0-value). But with seperate channels i can fix my issue on the backend. Thanks.

**Reply by barbudor** [reply]
Innocent question : if you are using the Shelly EM that is single channel, you shouldn't see energy for more than 1 phase/channel.
Are you sure your device is correctly configured for the [EM](https://templates.blakadder.com/shelly_EM.html) ?

**Reply by sfromis** [reply]
The [Shelly EM](https://shelly.cloud/knowledge-base/devices/shelly-em/) hardware does have two CT clamps.

**Reply by barbudor** [reply]
AH indeed. You need to go to the wiring diagram to find this. Nowhere in the production description it says 2 energy channels

**Reply by sfromis** [reply]
Illustration from Shelly: 
![image](https://user-images.githubusercontent.com/47082390/162567280-03b03dc2-1df1-4a0b-a49c-c7140bbd73e2.png)

**Reply by abroeders** [reply]
This diagram is exactly my situation. My initial problem was that 10KWh consumption and 12KWh production would sum up to 22KWh. I expected a total of -2KWh.
With 'SetOption129 1' I now have 2 counters, one for production and one for consumption.

**Reply by barbudor** [reply]
You should be able to perform the substraction on Tasmota using rules and publish that in addition to the separated energies
I would always consider that it is better to keep separed pieces of information separated as longer as possible instead of combining them at the source and loosing information

**Reply by barbudor** [reply]
The reason for SetOption129 is that originally the Tasmota Energy model was single channel
So when devices with more than 1 channel appeared, in order not to break the existing, it was implemented as the sum of all channels
Then as it appears there was a strong intrerrest for separated information, the 3 phases model was added but the SetOption129 was introduces so people would willingly switch to that model and avoid breaking change

**Reply by abroeders** [reply]
Other related question. How to use PowerSet/PowerCal with a two channel device? Do both channels always have the same calibration? What's the difference between PowerSet and PowerCal?

**Reply by barbudor** [reply]
Yes, hte same calibration is applied to both channels. It may or may not be exact depending on the device.
PowerSet set the calibration based on providing a power value measured externally. A calibration value is computed from this.
PowerCal provide the directly the calibration value. Could be used to save the calibration value and restore it later with rdoing the calibration with a know power load.

**Comment by hagenbuch** [comment]
Maybe it has been dealt with already, but I have to add my 2 cents here:

As far as momentaneous production or consumption (W) is concerned, negative values certainly are "good enough" (BTW just like VAr)

However, when dealing with an amount of energy over a certain time, always two integral counters should be provided: One for consumption in Wh and one for production in Wh.

The integrals should be filled correctly according to the "direction" in quasi every cycle: The power could run "forward" for 100ms and then "backward" for another 500ms. This should be reflected in both counters for this is exactly what an offical modern counter would do.

It's even "worse" for triphase counters: In total, Tasmota would need six counters (if we have to deal with reactive power maybe 12!) for it can always be that on one phase, power is "consumed" and on the other two "produced". This well get more and more important with upcoming "balcony solar installations", where phases will be unbalanced even into the negative.

Andreas
